app/models/user/permissions.rb

```diff
@@ -39,27 +40,53 @@ def admin_sections=(sections)
     keep_permissions = existing_permissions.values_at(*keep)
     new_permissions = create.map { |section| member_permissions.build(admin_section: section) }
 
-    self.member_permissions = keep_permissions + new_permissions
+    all_permissions = keep_permissions + new_permissions
+    # keep service permissions as they were
+    all_permissions << services_member_permission if services_member_permission
+
+    self.member_permissions = all_permissions
   ensure
     @_admin_sections = nil
   end
 
-  def member_permission_service_ids=(service_ids)
-    services_section = [:services]
+  # returns all permissions (:portal, :finance, :settings, :partners, :monitoring, :plans) for admins
+  # and the allowed ones for member users
+  def member_permission_ids
+    admin? ? AdminSection.permissions : admin_sections - [:services]
+  end
 
+  def member_permission_ids=(roles)
+    self.admin_sections = Array(roles).reject(&:blank?)
+  end
+
+  def member_permission_service_ids=(service_ids)
     if service_ids.present?
-      self.admin_sections = admin_sections | services_section
-      services_member_permission.service_ids = service_ids
-    else
-      self.admin_sections = admin_sections - services_section
+      service_ids = Array(service_ids).reject(&:blank?).map(&:to_i)
+      if services_member_permission
+        services_member_permission.service_ids = service_ids & existing_service_ids
+        services_member_permission.save
+      else
+        new_services_member_permissions = member_permissions.build( admin_section: :services,
+                                                                      service_ids: service_ids)
+        member_permissions << new_services_member_permissions
+      end
+    elsif services_member_permission
+      self.member_permissions = member_permissions - [services_member_permission]
     end
+  ensure
+    @_admin_sections = nil
+  end
+
+  def existing_service_ids
+    account.try(:service_ids) || []
   end
 
   # TODO: this is suboptimal to do on 'read', it should be done on write
   # but then it is much harder to test MemberPermission#service_ids=
+  # returns [] if no services are enabled, and nil if all (current and future) services are enabled
   def member_permission_service_ids
+    return nil if admin? || !services_member_permission
```

Comment:
This is needed to distinguish between _"All current and future APIs"_ available (`nil`) and no APIs available (all checkboxes are empty) (`[]`).

Reply:
Shouldn't the return statement be at the end of the method?

Reply:
Look at [RuboCop's GuardClause](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/GuardClause).

Although I personally agree with the _GuardClause_ that with this return at the beginning is easier to understand, I know that this something very subjective and not everyone agrees 😄

Reply:
Yes I like using guards too, but this one is somewhat weird, probably because of ruby:
`return nil if admin?` -> OK if admin?, then return nil, otherwise continue. This is a standard guard AFAIK.
However the `|| !services_member_permission` is making the return to always return and not continue, am I wrong? _return `nil` id admin, otherwise return `!services_member_permission`_.

Reply:
@josemigallas Imagine it as: `return nil if (admin? || !services_member_permission)` 😄 
It is doing: return nil is it is an admin user or if it doesn't have a services_member_permission 😄

Reply:
@josemigallas Yes, it's what @Martouta saying: return `nil` straight away if either of the two conditions holds: `admin?` meaning the user is admin OR `!services_member_permission` meaning that there is no `MemberPermission` with `admin_section: services` meaning that all services are enabled (see above).